File: 2sem-met3/py/test_met3.py
import pytest
import numpy as np
import matplotlib.pyplot as plt
import met3


def test_met3():
    print(f'running met3 test')
    N = 5
    dim = 3
    M = 101

    x0 = np.linspace(-1, 1, N)
    x1 = np.linspace(-1, 1, M)

    y0 = met3.func(x0)  # here's your function (change met3.func())
    y1 = met3.func(x1)

    # here's your approximation (change met3.approx())
    y_algpoly, _, p_algpoly = met3.approx(x0, y0, x1, met3.ApproxType.algebraic, dim)
    y_legpoly, _, p_legpoly = met3.approx(x0, y0, x1, met3.ApproxType.legendre, dim)

    assert(all(abs(y1 - y_algpoly) < 1))
    assert(all(abs(y1 - y_legpoly) < 1))
    assert(len(p_algpoly) == dim)
    assert(len(p_legpoly) == dim)
    # assert(all(abs(p_algpoly - p_legpoly) < 1e-3))

    plt.figure(1)
    plt.title('Y(X)')
    plt.plot(x1, y1, 'ko', label='exact')
    plt.plot(x1, y_algpoly, 'b-p', label='algebraic')
    plt.plot(x1, y_legpoly, 'g:*', label='legendre')
    plt.legend()

    plt.figure(2)
    plt.title('log error')
    plt.plot(x1, -np.log10(np.abs(y1 - y_algpoly)), 'b-p', label='algebraic')
    plt.plot(x1, -np.log10(np.abs(y1 - y_legpoly)), 'g:*', label='legendre')
    plt.legend()
    plt.show()

File: 2sem-met3/matlab/test_met3.m
clear all
close all
clc

%%
N = 5;
x0 = linspace(-1, 1, N)';
dim = 3;

M = 101;
x1 = linspace(-1, 1, M)';

y0 = met3_func(x0);  % here's your function (change met3_func.m)
y1 = met3_func(x1);

%%
% here's your approximation (change met3_approximate.m)
[y_algpoly, ~, P_algpoly] = met3_approximate(x0, y0, x1, 0, dim);
assert(numel(P_algpoly) == dim, ...
        ['polynome length should be ' num2str(dim)])
assert(all(abs(y1 - y_algpoly) < 1), ...
        'algebraic polynome approximation is too bad')

[y_legpoly, ~, P_legpoly] = met3_approximate(x0, y0, x1, 1, dim);
assert(numel(P_legpoly) == dim, ...
        ['polynome length should be ' num2str(dim)])
assert(all(abs(y1 - y_legpoly) < 1), ...
        'legendre polynome approximation is too bad')
% assert(all(abs(P_algpoly - P_legpoly) < 1e-3))

%%
figure(1), hold on, title('log error')
plot(x1, -log10(abs(y1 - y_algpoly)), 'b-p')
plot(x1, -log10(abs(y1 - y_legpoly)), 'g:*')

figure(2), hold on, title('Y(X)')
plot(x1, y1, 'ko')
plot(x1, y_algpoly, 'b-p')
plot(x1, y_legpoly, 'g:*')
